R/input-utils.R

```diff
@@ -2,30 +2,62 @@ controlLabel <- function(controlName, label) {
   label %AND% tags$label(class = "control-label", `for` = controlName, label)
 }
 
+# This function takes in either a list or vector for `choices` (and
+# `choiceNames` and `choiceValues` are passed in as NULL) OR it takes
+# in a list or vector for both `choiceNames` and `choiceValues` (and
+# `choices` is passed as NULL) and returns a list of two elements:
+#    - `choiceNames` is a vector or list that holds the options names
+#      (each element can be arbitrary UI, or simple text)
+#    - `choiceValues` is a vector or list that holds the options values
+#       (each element must be simple text)
+normalizeChoicesArgs <- function(choices, choiceNames, choiceValues) {
+  # if-else to check that either choices OR (choiceNames + choiceValues)
+  # were correctly provided
+  if (is.null(choices)) {
+    if (length(choiceNames) == 0 || length(choiceValues) == 0) {
+      stop("Please specify a non-empty vector for `choices` (or,
+           alternatively, for both `choiceNames` and `choiceValues`).")
+    }
+    if (length(choiceNames) != length(choiceValues)) {
+      stop("`choiceNames` and `choiceValues` must have the same length.")
+    }
+    if (anyNamed(choiceNames) || anyNamed(choiceValues)) {
+      stop("`choiceNames` and `choiceValues` must not be named.")
+    }
+  } else {
+    if (!is.null(choiceNames) || !is.null(choiceValues)) {
+      warning("Using `choices` argument; ignoring `choiceNames` and `choiceValues`.")
+    }
+    choices <- choicesWithNames(choices) # resolve names if not specified
+    choiceNames <- names(choices)
+    choiceValues <- unname(choices)
+  }
+
+  return(list(choiceNames = choiceNames, choiceValues = choiceValues))
+}
 
 # Before shiny 0.9, `selected` refers to names/labels of `choices`; now it
 # refers to values. Below is a function for backward compatibility. It also
 # coerces the value to `character`.
-validateSelected <- function(selected, choices, inputId) {
+normalizeSelected <- function(selected, inputId, choiceNames, choiceValues) {
```

Comment:
Since 0.10.0 (which was released in June 2014) users have gotten a warning if `selected` was a name instead of a value. So I think it's safe to assume that no one is doing that anymore. And that means that we can remove this function entirely. We will need to do `selected <- as.character(selected)` where we previously called `validateSelected`, for the reasons in the comment directly below.